crates/gitbutler-tauri/tauri.conf.json

```diff
@@ -13,7 +13,7 @@
 		"allowlist": {
 			"fs": {
 				"readFile": true,
-				"scope": ["$APPCACHE/archives/*", "$RESOURCE/_up_/scripts/*"]
+				"scope": ["$APPCACHE/archives/*", "$RESOURCE/_up_/scripts/*", "$HOME/**"]
```

Comment:
Hmm, I wonder if there is a way to not give the frontend javascript code access to `$HOME`, from a security standpoint

Reply:
Following up on the above, to fetch local images using `convertFileSrc`, the linked issue suggests using "**". Since I couldn’t find a better reference, I referred to [tauri docs](https://tauri.app/v1/api/js/fs/#security) and decided to use the `$HOME` directory near the project’s path.
Currently, the Tauri configuration only allows directory read access, which seemed acceptable at the time.

I suggest implementing protection against directory traversal attacks and unintended path access by validating paths through the validateFilePath function. If you have any better alternatives or suggestions, I would appreciate it if you could let me know.

**fileDiff.svelte**

```ts 

import { getFileExtension, validateFilePath } from '$lib/utils/filePath';

onMount(async () => {
    if (isBinary || !isLarge) {
      const validatedPath = await validateFilePath(project.path, filePath);
      if (validatedPath) {
        fileSize = await getBinFileSize(validatedPath);
        if (isImage) {
          try {
            imageSrc = convertFileSrc(validatedPath);
          } catch (error) {
            console.error('Failed to convert image path:', error);
            console.error('Attempted image source:', validatedPath);
          }
        }
      } else {
        console.error('Invalid file path:', filePath);
        fileSize = 'Invalid path';
      }
    }
  });
```

**utils/filePath.ts**

```ts
import { normalize, isAbsolute, join, homeDir } from '@tauri-apps/api/path';

//...

export async function validateFilePath(projectPath: string, filePath: string): Promise<string | false> {
  try {
    const normalizedProjectPath = await normalize(projectPath);
    const fullPath = await join(normalizedProjectPath, filePath);
    const normalizedFullPath = await normalize(fullPath);
    const userHomeDir = await homeDir();

    // Check if it's an absolute path
    if (!(await isAbsolute(normalizedFullPath))) {
      return false;
    }

    // Check if the path is within the project directory
    if (!normalizedFullPath.startsWith(normalizedProjectPath) && !normalizedFullPath.startsWith(userHomeDir))  {
      return false;
    }

    // Check file extension
    const allowedExtensions = ['txt', 'pdf', 'jpg', 'png', 'gif', 'svg', 'webp', 'ttf'];
    const extension = getFileExtension(filePath);
    if (extension && !allowedExtensions.includes(extension)) {
      return false;
    }

    return normalizedFullPath; // Return the normalized full path if valid
  } catch (error) {
    console.error('Error validating file path:', error);
    return false;
  }
}
```

Reply:
Hey yeah so we have a very similar use-case that we're working on in [this PR](https://github.com/gitbutlerapp/gitbutler/pull/4850). We want to get the contents of every pull request template file from the local repo / project dir. 

It's not quite ready, but the plan is to add a rust function we'll expose to the FE to `invoke` which will accept a relative path. Relative to the project (repositories) root path. Well then use the built-in git cmds to `cat` the target file. This way the FE can't access anything outside of the project / repo root directory and by leveraging the git cat-file, we'll also effectively disallow access to the `.git` dir and also any `.gitignore`d files. 

After talking it over with the team a bit, we figured this is a pretty safe way to expose a readFile fn to the FE.

Reply:
I love the outcome!
Just to share my thoughts, I think this will work as GB effectively adds everything that it sees in the working tree, minus large files, to the repository right away to create virtual-branch trees, which makes everything available as Git blobs as well which are referenced from these trees.

Personally, I hope that one day this can be changed to only do this when absolutely necessary or even never *implicitly* do it at all. By that time, there probably must be a way to fetch files directly from the worktree, and as Git has to protect itself from similar abuse, there is [a utility](https://docs.rs/gix-validate/latest/gix_validate/path/fn.component.html) to prevent. It's quite exhaustive as well.